match/allegations_clusters_nopd.py

```diff
@@ -0,0 +1,41 @@
+import sys
+sys.path.append('../')
+import pandas as pd
+from lib.path import data_file_path
+from datamatch import ThresholdMatcher, ColumnsIndex, StringSimilarity
+import numpy
+
+
+def cluster_officers_by_tracking_number(cprr):
+    df = cprr[['uid', 'first_name', 'last_name', 'tracking_number']]
+    df = df.drop_duplicates(subset='uid', keep='first').set_index('uid')
+    df.loc[:, 'tn'] = df.tracking_number.fillna('').map(lambda x: x[:])
+    matcher = ThresholdMatcher(ColumnsIndex('tn'), {
+        'tracking_number': StringSimilarity()
+    }, df, show_progress=True)
```

Comment:
so it looks like `tn` column is almost exactly the same as `tracking_number`? Well, `ColumnsIndex` already ensures that rows with different `tn` do not get to match with each other so what is likely to happen is you only see exact matches (score is never below 1). Is that the intention? If it is then something like this is both shorter and run faster:
```python
df = df.set_index('tracking_number')
clusters = dict()
for tn, row in df.iterrows():
    clusters.setdefault(tn, list()).append(row)
# discard tracking numbers that involve less than 2 officers
clusters = {tn: l for tn, l in clusters.items() if len(l) > 1}

Reply:
That is the intention. Thanks for the feedback!

Reply:
Sorry, even my first attempt is the wrong way to do it ;). It could even be shorter and faster:
```python
grouped = cprr[['uid', 'first_name', 'last_name', 'tracking_number']]\
    .groupby("tracking_number").filter(lambda x: len(x) > 1)
```
I'm sure whatever else you want to do can be done with just "group by" operations. Can you read [this article](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html#filtration) and try to do everything with just operations discussed there?

Reply:
Beautiful. I initially started with a "group by" operation as I've been doing all other analysis with them, but I didn't think to filter with a lambda function. Awesome. Thank you!